search/views.py

```diff
@@ -318,12 +121,109 @@ def search(request):
+    url_query_params_string = re.sub(r"(&cluster_id=[0-9]*)", "", url_query_params_string)
+
+    result = cluster_sound_results(request, features=DEFAULT_FEATURES)
+
+    # check if computation is finished. If not, send computation state.
+    if result['finished']:
+        if result['result'] is not None:
+            results = result['result']
+            num_clusters = num_clusters = len(results) + 1
+        else:
+             return JsonResponse({'status': 'failed'}, safe=False)
+    elif result['error']:
+        return JsonResponse({'status': 'failed'}, safe=False)
+    else:
+        return JsonResponse({'status': 'pending'}, safe=False)
+
+    num_sounds_per_cluster = [len(cluster) for cluster in results]
+    classes = {sound_id: cluster_id for cluster_id, cluster in enumerate(results) for sound_id in cluster}
+
+    # label clusters using most occuring tags
+    sound_instances = sounds.models.Sound.objects.bulk_query_id(map(int, classes.keys()))
+    sound_tags = {sound.id: sound.get_sound_tags() for sound in sound_instances}
+    cluster_tags = defaultdict(list)
+    query_terms = {t.lower() for t in request.GET.get('q', '').split(' ')}
+    for sound_id, tags in sound_tags.iteritems():
+        cluster_tags[classes[str(sound_id)]] += [t.lower() for t in tags if t.lower() not in query_terms]
+    cluster_tags_with_count = {k: sorted([(t, tags.count(t)) for t in set(tags)], 
+                                         key=lambda x: x[1], reverse=True)
+                               for k, tags in cluster_tags.iteritems()}
+    cluster_most_occuring_tags = [' '.join(zip(*tags[:3])[0]) for tags in cluster_tags_with_count.values() if len(tags)>2]  # dict values sorted?!
+
+    return render(request, 'search/clustering_facet.html', {
+            'results': classes,
+            'url_query_params_string': url_query_params_string,
+            'cluster_id_num_results': zip(range(num_clusters), num_sounds_per_cluster, cluster_most_occuring_tags),
+    })
+
+
+def cluster_visualisation(request):
+    url_query_params_string = request.META['QUERY_STRING']
+    return render(request, 'search/clustered_graph.html', {
+            'url_query_params_string': url_query_params_string,
+    })
+
+
+def clustered_graph(request):
+    """Returns the clustered sound graph representation of the search results.
+    """
+    result = cluster_sound_results(request, features=DEFAULT_FEATURES)
+    graph = result['graph']
+
+    results = sounds.models.Sound.objects.bulk_query_id([int(node['id']) for node in graph['nodes']])
+    preview_urls_name_tags_by_id = {s.id:(s.get_preview_abs_url()[21:],
+                                          s.original_filename,
+                                          ' '.join(s.get_sound_tags()),
+                                          s.get_absolute_url()) for s in results}
+
+    for node in graph['nodes']:
+        node['url'] = preview_urls_name_tags_by_id[int(node['id'])][0]
```

Comment:
Maybe `preview_urls_name_tags_by_id` could be renamed to something like `sound_metadata`?